src/rosdiscover/observer/ros1_actions.py

```diff
@@ -0,0 +1,270 @@
+    result: Optional[Topic] = attr.ib(default=None)
+
+    @property
+    def fmt(self) -> str:
+        assert self.is_complete()
+        assert self.goal is not None  # Needed for mypy
+        match = _GOAL_FMT.match(self.goal.format)
+        assert match is not None
+        return match.group(1)
+
+    def is_complete(self) -> bool:
+        """Check if all information is there for an action.
+
+        Returns
+        -------
+        bool
+            Whether the action is complete
+        """
+        has_goal = self.goal is not None
+        has_cancel = self.cancel is not None
+        has_status = self.status is not None
+        has_feedback = self.feedback is not None
+        has_result = self.result is not None
+        return has_goal and has_cancel and has_status and has_feedback and has_result
+
+    def add(self,
+            context: NodeContext,
+            topic: Topic,
+            sub_if_server: bool) -> None:
+        """Add the topic back into the context.
+
+        Parameters
+        ----------
+        context: NodeContext
+            The node context to add the topic to
+        topic: Topic
+            The topic to add to the context
+        sub_if_server: bool
+            Indicate wheter to subscribe or publish if this was expected to be a server
+            or a client
+        """
+        if self.is_server:  # i.e., the candidate is an action server
+            if sub_if_server:  # i.e., action servers subscribe to this topic
+                context.sub(topic.name, topic.format)
+            else:  # i.e., action servers publish this topic
+                context.pub(topic.name, topic.format)
+        else:  # i.e., the candidate is an action client
+            if sub_if_server:  # i.e., action clients publish this topic
+                context.pub(topic.name, topic.format)
+            else:  # i.e., action clients subscribe to this topic
+                context.sub(topic.name, topic.format)
+
+    def add_unfinished_to_context(self, nc: NodeContext) -> None:
+        """Add topics from incomplete action server/client to node context
+
+        Parameters
+        ----------
+        nc: NodeContext
+            The node to add the topics to
+        """
+        assert not self.is_complete()
+        if self.goal is not None:
+            self.add(nc, self.goal, True)
+        if self.cancel is not None:
+            self.add(nc, self.cancel, True)
+        if self.status is not None:
+            self.add(nc, self.status, False)
+        if self.feedback is not None:
+            self.add(nc, self.feedback, False)
+        if self.result is not None:
+            self.add(nc, self.result, False)
+
+
+def get_action_candidate(node: str,
+                         action_name: str,
+                         existing: Dict[str, Dict[str, ActionCandidate]],
+                         is_server: bool) -> ActionCandidate:
+    """Gets the action candidate if it exsits
```

Comment:
Small typo: `exsits` -> `exists`.
This method also seems to create a new `ActionCandidate` if none exists rather than just trying to fetch one?